### PR TITLE
bpo-43651: Fix EncodingWarning in test_io

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue43651

### CODE REVIEW

File: Lib/_pyio.py [hunk 1]
```diff
@@ -324,6 +324,8 @@ class OpenWrapper:
     __doc__ = DocDescriptor()
 
     def __new__(cls, *args, **kwargs):
+        if len(args) >= 2 and "b" not in args[1]:
+            kwargs["encoding"] = text_encoding(kwargs.get("encoding"))
```
Comment (vstinner) [inline Lib/_pyio.py]: Why does _io emit the warning in _io.TextIOWrapper, but you chose to modify OpenWrapper in _pyio?

You might extract this fix into a separated PR.

Comment (methane) [reply Lib/_pyio.py]: Because EncodingWarning was emitted for the OpenWrapper, not for the caller.

Comment (methane) [reply Lib/_pyio.py]: There was a bug in this `if` statement. I revert this.

**Comment by methane** [comment]
This PR can be simplified after #25103 is merged.

**Comment by methane** [comment]
#25103 did not help OpenWrapper(). I don't think we should allow `encoding="locale"` in binmode only for test_io.

File: Lib/test/test_io.py [hunk 2]
```diff
@@ -407,10 +407,12 @@ def large_file_ops(self, f):
     def test_invalid_operations(self):
         # Try writing on a file opened in read mode and vice-versa.
         exc = self.UnsupportedOperation
-        for mode in ("w", "wb"):
-            with self.open(os_helper.TESTFN, mode) as fp:
-                self.assertRaises(exc, fp.read)
-                self.assertRaises(exc, fp.readline)
+        with self.open(os_helper.TESTFN, "w", encoding="ascii") as fp:
```
Comment (vstinner) [inline Lib/test/test_io.py]: I would prefer to replace all "ascii" with "utf-8". Come on, ASCII is now legacy!

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**